samples/build-duration-logger/index.html

```diff
@@ -7,31 +7,65 @@
 
 <body>
 <script>
-    const gradleEnterpriseServer = 'https://gradle.my-company.com'; // replace with your Gradle Enterprise base url
-    const buildStreamUrl = `${gradleEnterpriseServer}/build-export/v1/builds/since/now?stream`;
+    const gradleEnterpriseServerUrl = 'https://gradle.my-company.com'; // replace with your Gradle Enterprise base url
+    const buildStreamUrl = `${gradleEnterpriseServerUrl}/build-export/v1/builds/since/now?stream`;
     const buildStream = new EventSource(buildStreamUrl);
-    const buildEventStreamUrl = buildId => `${gradleEnterpriseServer}/build-export/v1/build/${buildId}/events?eventTypes=BuildStarted,BuildFinished`;
+
+    class BuildDurationProcessor {
+
+        constructor(gradleEnterpriseServerUrl, maxConcurrentBuildsToProcess = 6) {
+            this.gradleEnterpriseServerUrl = gradleEnterpriseServerUrl;
+            this.pendingBuildIds = [];
+            this.numBuildsInProcess = 0;
+            this.maxConcurrentBuildsToProcess = maxConcurrentBuildsToProcess;
+        }
+
+        enqueue(buildId) {
+            this.pendingBuildIds.push(buildId);
+            this.processPendingBuilds();
+        }
+
+        processPendingBuilds() {
+            if (this.pendingBuildIds.length > 0 && this.numBuildsInProcess < this.maxConcurrentBuildsToProcess) {
+                this.processBuild(this.pendingBuildIds.shift());
+                setTimeout(() => this.processPendingBuilds(), 0);
+            }
+        }
+
+        createBuildDurationEventStreamUrl(buildId) {
+            return `${this.gradleEnterpriseServerUrl}/build-export/v1/build/${buildId}/events?eventTypes=BuildStarted,BuildFinished`;
+        }
+
+        processBuild(buildId) {
+            this.numBuildsInProcess++;
+            const buildDurationEventStream = new EventSource(this.createBuildDurationEventStreamUrl(buildId));
+
+            let startTime;
+
+            buildDurationEventStream.addEventListener('BuildEvent', event => {
+                const buildEventPayload = JSON.parse(event.data);
+                const eventType = buildEventPayload.type.eventType;
+
+                if (eventType == 'BuildStarted') {
+                    startTime = buildEventPayload.timestamp;
+                } else if (eventType == 'BuildFinished') {
+                    const endTime = buildEventPayload.timestamp;
+
+                    console.log(`Build ${buildId} completed in ${endTime - startTime}ms`);
+                    this.numBuildsInProcess--;
+                }
+            });
+        }
+    };
+
+    const buildDurationProcessor = new BuildDurationProcessor(gradleEnterpriseServerUrl);
```

Comment:
What if we move the consts above here instead of having some above and the rest below? I don't like that the class needs to be defined before but I think that it is easier to understand how everything is built and wired if one sees all the consts together.

Reply:
Done